File: Typecast/TCGlyphPathFactory.m
//
//  TCGlyphPathFactory.m
//  Type Designer
//
//  Created by David Schweinsberg on 21/12/12.
//  Copyright (c) 2012 David Schweinsberg. All rights reserved.
//

#import "TCGlyphPathFactory.h"
#import "TCGlyph.h"
#import "TCPoint.h"
#import <ApplicationServices/ApplicationServices.h>

static int midValue(int a, int b)
{
    return a + (b - a) / 2;
}

static void addContourToPath(CGMutablePathRef path, TCGlyph *glyph, int startIndex, int count)
{
    int offset = 0;
    while (offset < count)
    {
        //TCPoint *point_minus1 = [glyph points][(offset == 0) ? startIndex + count - 1 : startIndex + (offset - 1) % count];
        TCPoint *point = [glyph points][startIndex + offset % count];
        TCPoint *point_plus1 = [glyph points][startIndex + (offset + 1) % count];
        TCPoint *point_plus2 = [glyph points][startIndex + (offset + 2) % count];

        if (offset == 0)
            CGPathMoveToPoint(path, NULL, [point x], [point y]);

        if (point.onCurve && point_plus1.onCurve)
        {
            CGPathAddLineToPoint(path, NULL, [point_plus1 x], [point_plus1 y]);
            offset++;
        }
        else if (point.onCurve && !point_plus1.onCurve && point_plus2.onCurve)
        {
            CGPathAddQuadCurveToPoint(path, NULL, [point_plus1 x], [point_plus1 y], [point_plus2 x], [point_plus2 y]);
            offset+=2;
        }
        else if (point.onCurve && !point_plus1.onCurve && !point_plus2.onCurve)
        {
            CGPathAddQuadCurveToPoint(path, NULL, [point_plus1 x], [point_plus1 y], midValue([point_plus1 x], [point_plus2 x]), midValue([point_plus1 y], [point_plus2 y]));
            offset+=2;
        }
        else if (!point.onCurve && !point_plus1.onCurve)
        {
            CGPathAddQuadCurveToPoint(path, NULL, [point x], [point y], midValue([point x], [point_plus1 x]), midValue([point y], [point_plus1 y]));
            offset++;
        }
        else if (!point.onCurve && point_plus1.onCurve)
        {
            CGPathAddQuadCurveToPoint(path, NULL, [point x], [point y], [point_plus1 x], [point_plus1 y]);
            offset++;
        }
        else
        {
            NSLog(@"drawGlyph case not catered for!!");
        }
    }
}

@implementation TCGlyphPathFactory

+ (CGPathRef)buildPathWithGlyph:(TCGlyph *)glyph
{
//    if (glyph == nil)
//        return nil;

    CGMutablePathRef glyphPath = CGPathCreateMutable();

    // Iterate through all of the points in the glyph.  Each time we find a
    // contour end point, add the point range to the path.
    int firstIndex = 0;
    int count = 0;
    for (int i = 0; i < [[glyph points] count]; ++i)
    {
        count++;
        if ([[glyph points][i] isEndOfContour]) {
            addContourToPath(glyphPath, glyph, firstIndex, count);
            firstIndex = i + 1;
            count = 0;
        }
    }
    return glyphPath;
}

@end

File: Typecast/TCGlyphPathFactory.swift
//
//  TCGlyphPathFactory.swift
//  Type Designer
//
//  Created by David Schweinsberg on 7/25/17.
//  Copyright © 2017 David Schweinsberg. All rights reserved.
//

import Foundation
import os.log

class TCGlyphPathFactory {

  class func midValue(_ a: Int, _ b: Int) -> Int {
    return a + (b - a) / 2
  }

  class func addContourToPath(path: CGMutablePath, glyph: TCGlyph, startIndex: Int, count: Int) {
    var offset = 0
    while offset < count {
      let point = glyph.points[startIndex + offset % count]
      let point_plus1 = glyph.points[startIndex + (offset + 1) % count]
      let point_plus2 = glyph.points[startIndex + (offset + 2) % count]

      if offset == 0 {
        path.move(to: CGPoint(x: point.x, y: point.y))
      }

      if point.onCurve && point_plus1.onCurve {
        path.addLine(to: CGPoint(x: point_plus1.x, y: point_plus1.y))
        offset += 1
      } else if point.onCurve && !point_plus1.onCurve && point_plus2.onCurve {
        path.addQuadCurve(to: CGPoint(x: point_plus2.x, y: point_plus2.y),
                          control: CGPoint(x: point_plus1.x, y: point_plus1.y))
        offset += 2
      } else if point.onCurve && !point_plus1.onCurve && !point_plus2.onCurve {
        path.addQuadCurve(to: CGPoint(x: midValue(point_plus1.x, point_plus2.x),
                                      y: midValue(point_plus1.y, point_plus2.y)),
                          control: CGPoint(x: point_plus1.x, y: point_plus1.y))
        offset += 2
      } else if !point.onCurve && !point_plus1.onCurve  {
        path.addQuadCurve(to: CGPoint(x: midValue(point.x, point_plus1.x),
                                      y: midValue(point.y, point_plus1.y)),
                          control: CGPoint(x: point.x, y: point.y))
        offset += 1
      } else if !point.onCurve && point_plus1.onCurve {
        path.addQuadCurve(to: CGPoint(x: point_plus1.x, y: point_plus1.y),
                          control: CGPoint(x: point.x, y: point.y))
        offset += 1
      } else {
        os_log("drawGlyph case not catered for!!")
      }
    }
  }

  class func buildPath(with glyph: TCGlyph) -> CGPath {

    let glyphPath = CGMutablePath()

    // Iterate through all of the points in the glyph.  Each time we find a
    // contour end point, add the point range to the path.
    var firstIndex = 0
    var count = 0
    for i in 0 ..< glyph.points.count {
      count += 1
      if glyph.points[i].endOfContour {
        addContourToPath(path: glyphPath, glyph: glyph, startIndex: firstIndex,
                         count: count)
        firstIndex = i + 1
        count = 0;
      }
    }
    return glyphPath
  }
}
